### PR TITLE
add more examples to syntax highlighting tutorial

### PR DESCRIPTION
**Pull request by takase1121**

### CODE REVIEW

File: locales/en/tutorials/syntax-highlighting.md [hunk 1]
```diff
@@ -178,3 +179,273 @@ To test your new syntax highlighting you need to do two things:
...
+This is very obvious when the user tries to resize the editor with `ctrl-scroll` or `ctrl+` and `ctrl-`.
+Please use it in moderation.
+
+### Space handling (v2.1 (upcoming) / `master`)
+
+By default, Lite XL prepends a pattern `{ pattern = "%s+", type = "normal" }` to the syntax.
+This improves the performance drastically on lines that starts with whitespaces (eg. heavily indented lines)
+by matching the whitespace before other patterns in order to prevent Lite XL from iterating the entire syntax.
+However, there may be syntaxes that require matching spaces (eg. Markdown with indented blocks) so this can be disabled by
+setting `space_handling` to `false.`
```
Comment (jgmdev) [inline locales/en/tutorials/syntax-highlighting.md]: Maybe add additional comment saying something like "to keep the space handing optimization on your syntax file add the rule `%s+` after the patterns that require spaces"

**Comment by jgmdev** [comment]
Looks good to me, one thing which comes to mind and may be good to know is that when doing group matching, if one wants the matched items to get colorized as specified on the `symbols` table one should not include spaces to it, for example:

instead of doing something like:
```lua
{pattern = "[%a_][%w_]+%s+()[%a_][%w_]+", type = {"keyword2", "symbol"}}
```
one should do:
```lua
{pattern = "[%a_][%w_]+()%s+()[%a_][%w_]+", type = {"keyword2", "normal", "symbol"}}
```
This way the first match instead of using the specified type `keyword2` it would use the one from the symbols table if the matched element exists on the table.

Another thing that confuses is not knowing the fact that the patterns should include rules that are able to match elements from the `symbols` table, lets say for example:

you have a pattern like this at the end of the `patterns` table which is pretty common on syntax files:

```lua
{ pattern = "[%a_][%w_]*", "symbol" }
```
And you are adding to the symbols table things like this:
```lua
symbols = {
  ["my-symbol"] = "function",
  ["..something_else"] = "literal"
}
```

Since `my-symbol` has a dash and `..something_else` has dots it will never get matched if one doesn't adds some patterns like:

```lua
{ pattern = "[%a_][%w%-_]*", "symbol" },
{ pattern = "%.%.[%a_][%w_]*", "symbol" },
```
This is something that is not evident for new contributors as seen on this commit:

https://github.com/lite-xl/lite-xl-plugins/pull/107/commits/7423e088453bf6a483ad80d73f8421f393eaca21

**Comment by takase1121** [comment]
I don't understand this. 


> Another thing that confuses is not knowing the fact that the patterns should include rules that are able to match elements from the `symbols` table, lets say for example:
> 
> you have a pattern like this at the end of the `patterns` table which is pretty common on syntax files:
> 
> ```lua
> { pattern = "[%a_][%w_]*", "symbol" }
> ```
> 
> And you are adding to the symbols table things like this:
> 
> ```lua
> symbols = {
>   ["my-symbol"] = "function",
>   ["..something_else"] = "literal"
> }
> ```
> 
> Since `my-symbol` has a dash and `..something_else` has dots it will never get matched if one doesn't adds some patterns like:
> 
> ```lua
> { pattern = "[%a_][%w%-_]*", "symbol" },
> { pattern = "%.%.[%a_][%w_]*", "symbol" },
> ```
> 
> This is something that is not evident for new contributors as seen on this commit:
> 
> [lite-xl/lite-xl-plugins@7423e08](https://github.com/lite-xl/lite-xl-plugins/commit/7423e088453bf6a483ad80d73f8421f393eaca21)

Does this mean that  `{ pattern = "[%a_][%w_]*", "symbol" }` is added to every syntax by Lite XL?

> Looks good to me, one thing which comes to mind and may be good to know is that when doing group matching, if one wants the matched items to get colorized as specified on the `symbols` table one should not include spaces to it, for example:
> 
> instead of doing something like:
> 
> ```lua
> {pattern = "[%a_][%w_]+%s+()[%a_][%w_]+", type = {"keyword2", "symbol"}}
> ```
> 
> one should do:
> 
> ```lua
> {pattern = "[%a_][%w_]+()%s+()[%a_][%w_]+", type = {"keyword2", "normal", "symbol"}}
> ```
> 
> This way the first match instead of using the specified type `keyword2` it would use the one from the symbols table if the matched element exists on the table.
> 

Isn't the `()` supposed to mark the start of next token? Why would it interfere with the symbol?

**Comment by Guldoman** [comment]
> Does this mean that `{ pattern = "[%a_][%w_]*", "symbol" }` is added to every syntax by Lite XL?

No. Each language plugin should do as needed.

> Isn't the `()` supposed to mark the start of next token? Why would it interfere with the symbol?

The idea of the `symbols` table is this: each match done by the tokenizer is looked up in the `symbols` table.
Let's use `Lite XL` as the line we're matching, and `symbols = { ["Lite"] = "number" }` as the `symbols` table (so we expect that every time `Lite` is found, it's highlighted as `number`).
Using this
```Lua
{pattern = "[%a_][%w_]+%s+()[%a_][%w_]+", type = {"keyword2", "symbol"}}
```
We get:
1. `Lite ` as `keyword2`. Notice the space.
2. `XL` as `symbol`

Before adding this to the tokens list we first check if they're present in the `symbols` table.
* `symbols["Lite "] = nil`, so we add `Lite ` as `keyword2`;
* `symbols["XL"] = nil`, so we add `XL` as `symbol`.

Now using
```lua
{pattern = "[%a_][%w_]+()%s+()[%a_][%w_]+", type = {"keyword2", "normal", "symbol"}}
```
We get:
1. `Lite` as `keyword2`
2. ` ` as `normal`
3. `XL` as `symbol`

So we check the `symbols` table.
* `symbols["Lite"] = number`, so we add `Lite` as `number`;
* `symbols[" "] = nil`, so we add ` ` as `normal`;
* `symbols["XL"] = nil`, so we add `XL` as `symbol`.

**Comment by takase1121** [comment]
Added jgmdev's suggestion. The history is pretty messy, probably this PR should be squashed.

**Pull request merged**